examples/vqe.lisp

```diff
@@ -16,221 +16,300 @@
+    :initform (error-missing-initform :operators)
+    :documentation "Operators in the Hamiltonian")))
+
+(defun make-vqe-problem (&key (number-of-qubits *number-of-qubits*)
+                           (reference-state *reference-state*)
+                           (ansatz *ansatz*)
+                           (coefficients *coefficients*)
+                           (operators *operators*))
+  (flet ((coerce-to-adjustable-array (x)
+           (make-array (length x) :element-type (type-of (elt x 0)) :adjustable t :initial-contents x)))
+    (make-instance 'vqe-problem
+                   :number-of-qubits number-of-qubits
+                   :reference-state reference-state
+                   :ansatz ansatz
+                   :coefficients (coerce-to-adjustable-array coefficients)
+                   :operators (coerce-to-adjustable-array operators))))
+
+(defun make-ansatz-string (thetas &key dagger)
+  "Return Quil code that evaluates the values of the sequence THETAS and instantiates an ansatz (or its Hermitian conjugate, if DAGGER is T)."
+  (with-output-to-string (stream)
+    (loop :for i :below (length thetas)
+          :for theta := (elt thetas i)
+          :do (format stream "MOVE theta[~d] ~F~%" i theta))
+    (format stream "~:[REFERENCE~%ANSATZ~;DAGGER ANSATZ~%DAGGER REFERENCE~]~%" dagger)))
+
+(defun solve-vqe (vqe-problem &optional (initial-thetas *initial-thetas*))
   "Run the variational quantum eigensolver algorithm."
-  (flet ((objective-function (thetas)
-           "Compute the energy based on the values of THETAS."
-           (loop :with state-prep := (quil:parse-quil
-                                      (format nil "DECLARE theta REAL[2]~%~
-                                                   MOVE theta[0] ~F~%~
-                                                   MOVE theta[1] ~F~%~A~A"
-                                              (elt thetas 0) (elt thetas 1)
-                                              reference-state ansatz))
-                 :with expectations := (qvm-app::perform-expectation 'qvm-app::pure-state state-prep pauli-terms number-of-qubits)
-                 :for u :in pauli-coefficients :for v :in expectations :sum (* u v))))
-
-    (cl-grnm:grnm-optimize #'objective-function initial-thetas)))
+  (let ((number-of-qubits (slot-value vqe-problem 'number-of-qubits))
+        (reference-state (slot-value vqe-problem 'reference-state))
+        (ansatz (slot-value vqe-problem 'ansatz))
+        (operators (slot-value vqe-problem 'operators))
+        (coefficients (slot-value vqe-problem 'coefficients)))
+
+    (flet ((objective-function (thetas)
+             "Compute the energy based on the values of THETAS."
+             (loop :with list-of-operators := (coerce operators 'list)
+                   :with state-prep := (quil:parse-quil
+                                        (format nil "DECLARE theta REAL[2]~%~A~%~A~%~A~%" reference-state ansatz (make-ansatz-string thetas)))
```

Comment:
`(length thetas)` guaranteed to be 2 here?

Reply:
For this example, yes.